x-pack/auditbeat/cache/cache.go

```diff
@@ -0,0 +1,58 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cache
+
+// Cache is just a map being used as a cache.
+type Cache struct {
+	hashMap map[string]Cacheable
+}
+
+// Cacheable is the interface items stored in Cache need to implement.
+type Cacheable interface {
+	Hash() string
+}
+
+// New creates a new cache.
+func New() *Cache {
+	return &Cache{
+		hashMap: make(map[string]Cacheable),
+	}
+}
+
+// IsEmpty checks if the cache is empty.
+func (cache *Cache) IsEmpty() bool {
+	return len(cache.hashMap) == 0
+}
+
+// DiffAndUpdateCache takes a list of new items to cache, compares them to the current
+// cache contents, and returns both items new to the cache and items that are in the cache
+// but missing in the new data.
+func (cache *Cache) DiffAndUpdateCache(current []Cacheable) (new, missing []interface{}) {
+	// Check for and delete missing - what is no longer in current that was in the cache
+	for cacheKey, cacheValue := range cache.hashMap {
+		found := false
+		for _, currentValue := range current {
+			if currentValue.Hash() == cacheKey {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			missing = append(missing, cacheValue)
+			delete(cache.hashMap, cacheKey)
+		}
+	}
```

Comment:
`current` is not a map, so I would need to build a new map and add all `current` items and their hashes. Doing that every time we check the cache (e.g. every second) seems pretty expensive - most of the time nothing will have changed, or only one or two items. Am I missing something?

Reply:
Just wanted to loop back here. I won't push too hard on the O(n^2), since I'm a total Go noob, so I may be missing something in your code or in the Map initialization.

But my computer currently runs 500-ish processes. This means each time it's comparing the new list of processes with the last state, it's doing a multiple of 250 000 operations (500^2) instead of doing a multiple of 500 operations... This may be irrelevant, since typically servers run way less stuff than workstations.

So I just wanted to put that out there, but we can keep things as they are for now. This may be premature optimization.